dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -115,127 +118,217 @@ def _external_source_node_names(pipeline):
+        pipe_run_ref = proxy._schedule_batch(pipe_run_ref)
+    return DALIOutputBatchRef(pipe_run_ref, output_idx)
+
+
+# In-place modify `default_collate_fn_map` to handle DALIOutputSampleRef
+_default_collate_fn_map.update({DALIOutputSampleRef: _collate_dali_output_sample_ref_fn})
 
 
 class _DALIProxy:
-    def __init__(self, dali_input_names, dali_input_q, deterministic):
-        # External source instance names
-        self.dali_input_names = dali_input_names
+    class _WorkerData:
+        def __init__(self, worker_id):
+            self.worker_id = worker_id
+            self.pipe_run_ref = None
+            self.next_worker_batch_idx = 0
+            self.batch_sample_idx = 0
+
+    def __init__(self, signature, dali_input_q, deterministic):
         # If True, the request is not sent to DALI upon creation, so that it can be scheduled
         # always in the same order by the main process. This comes at a cost of performance
-        self.deterministic = deterministic
+        self._deterministic = deterministic
         # Shared queue with the server
-        self.dali_input_q = dali_input_q
-        # Torch worker id, to be filled on first call to worker_id()
-        self._worker_id = None
-        # Iteration index for the current worker
-        self.data_idx = 0
+        self._dali_input_q = dali_input_q
+        # Per worker
+        self._worker_data = None
+        # Override callable signature
+        self._signature = signature
+        self.__call__ = signature
+        # Current batch
+        self._curr_batch_params = {}
+        # get num outputs
+        self._num_outputs = self._get_num_outputs()
+
+    def _get_num_outputs(self):
+        return_annotation = self._signature.return_annotation
+        if isinstance(return_annotation, tuple):
+            return len(return_annotation)
+        elif return_annotation is Signature.empty:
+            return 0
+        else:
+            return 1
 
-    @property
-    def worker_id(self):
+    def _get_worker_id(self):
         """
         Getter for 'worker_id'. In case of torch data worker it is the worker info, and
         in case of a thread the thread identifier
         """
-        if self._worker_id is None:
-            worker_info = torchdata.get_worker_info()
-            self._worker_id = worker_info.id if worker_info else threading.get_ident()
-        return self._worker_id
+        worker_info = _torchdata.get_worker_info()
+        return worker_info.id if worker_info else threading.get_ident()
+
+    def _get_worker_data(self):
+        if self._worker_data is None:
+            self._worker_data = _DALIProxy._WorkerData(self._get_worker_id())
+        return self._worker_data
 
-    def _create_pipe_run_ref(self, inputs):
-        # Identifier of this request
-        batch_id = (self.worker_id, self.data_idx)
-        self.data_idx = self.data_idx + 1
-        return DALIPipelineRunRef(batch_id, inputs=inputs)
+    def _add_sample(self, bound):
```

Comment:
What does "bound" mean in this context? As a noun it has some meanings, but none that fits here.

Reply:
Will rename

Reply:
OK, I got what it _means_ but it's still probably not the best choice. I see that you don't really need the bound signature here - you need a dictionary. Perhaps it would be better to hide the signature stuff at call site and pass just the arguments here?